src/components/structures/TimelinePanel.tsx

```diff
@@ -1584,6 +1587,42 @@ class TimelinePanel extends React.Component<IProps, IState> {
             : null;
     }
 
+    /**
+     * Get a list of undecryptable events currently visible on-screen.
+     *
+     * @returns {MatrixEvent[] | null} A list of undecryptable events, or null if
+     *     the list of events could not be determined.
+     */
+    public getVisibleDecryptionFailures(): MatrixEvent[] | null {
+        const messagePanel = this.messagePanel.current;
+        if (!messagePanel) return null;
+
+        const messagePanelNode = ReactDOM.findDOMNode(messagePanel) as Element;
+        if (!messagePanelNode) return null; // sometimes this happens for fresh rooms/post-sync
+        const wrapperRect = messagePanelNode.getBoundingClientRect();
+        const screenTop = wrapperRect.top - VISIBLE_DECRYPTION_FAILURE_MARGIN;
+        const screenBottom = wrapperRect.bottom + VISIBLE_DECRYPTION_FAILURE_MARGIN;
+
+        let enteredVisibleRange = false;
+
+        const result: MatrixEvent[] = [];
+        for (const ev of this.state.liveEvents) {
```

Comment:
by the way: what is the difference between `state.events` and `state.liveEvents`?

Reply:
@duxovni any answers here?

Reply:
`events` includes pending events we've sent, which aren't relevant here.

Reply:
thanks. Please could you add this as a comment in `IState`.